language-design.md

```diff
@@ -74,7 +154,14 @@ implementation choices from users?
 
 **Response**
 
+We think that implementation choices can certainly affect the user experience. Consider a language in which user data is stored in a hash set, and users are allowed to define their own data types. In this case, we'd have 2 choices:
+1. Allow users to provide a hash function, implicitly revealing that we're storing their data in a hash set.
+2. Use some built-in hash function to try to hash custom types, even though the function may not be appropriate for the data present.
+In this case, one might want to choose the first option for performance reasons or the second to obscure implementation. Which option we end up choosing will affect the user experience in some way - in the first case, if in the future we moved away from hash sets all of the users' existing code would suddenly stop working (or we'd have to design a workaround). However, in the second case the user would have to deal with bad performance.
+
+When it comes to deciding whether to hide our implementation choices from users, there are (unsurprisingly) tradeoffs to consider. The tradeoffs to consider would probably vary by situation, and our choice on what to do will depend on our priorities when designing the language. 
 
+We can see examples of languages that choose to show or obscure their implementation choices in the real world. C++ chooses to expose the implementation of its sets (hash/tree), to allow users to optimize for performance. Python, by contrast, obscures the implementation of its dictionaries (which are really just hash maps) for the sake of simplicity. 
```

Comment:
Python obscures its choice to some extend, but not completely. You were still able to figure out that it uses hash maps, and users [can in fact](https://docs.python.org/2/reference/datamodel.html#object.__hash__) write their own `__hash__` functions.